### PR TITLE
Add aiobotocore instrumentation

### LINKED ISSUE #553: Add support for boto3
Currently, only boto has an instrumentor.
Unfortunately, boto is deprecated and boto3 is the AWS SDK library used.

We wish to add support for boto3 instrumentation as well.

### PR DESCRIPTION
**Pull request by dacevedo12**
# Description

Added instrumentor for aiobotocore, an async client for amazon services.

It's based on the sync botocore instrumentor, overriding the methods to patch the async codepaths

Fixes #553

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- I've been using it on production for a while with no problems.
- workflow tests should pass to be sure

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR: 
- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added
- [x] Documentation has been updated

### CODE REVIEW

**Comment by dacevedo12** [comment]
CI is all green in my branch now :heavy_check_mark:

**Review by nemoshlag — COMMENTED** [review]
LGTM but I got assigned due to docs/instrumentation ownership (which looks great). @mariojonke can you take a look? I believe you'll this is your domain 🙏

File: instrumentation/opentelemetry-instrumentation-aiobotocore/tests/test_aiobotocore_dynamodb.py
Comment (mariojonke) [inline instrumentation/opentelemetry-instrumentation-aiobotocore/tests/test_aiobotocore_dynamodb.py:1]: as far as i can see the tests are mostly duplicates of the ones in botocore. i'd imagine these tests and the ones in botocore will get out of sync pretty soon once some new functionality is added. 

is there maybe some way to unify them?

Comment (dacevedo12) [reply instrumentation/opentelemetry-instrumentation-aiobotocore/tests/test_aiobotocore_dynamodb.py:1]: I was thinking of using inheritance but it probably wouldn't be so elegant.

Do you know of any similar cases that could serve as inspiration?

It seems like ddtrace accepted that level of duplication in tests
- https://github.com/DataDog/dd-trace-py/blob/1.x/tests/contrib/aiobotocore/test.py
- https://github.com/DataDog/dd-trace-py/blob/1.x/tests/contrib/botocore/test.py

Comment (mariojonke) [reply instrumentation/opentelemetry-instrumentation-aiobotocore/tests/test_aiobotocore_dynamodb.py:1]: tbh, i haven't thought about it that deeply yet, but i was thinking in the direction of dynamically generating the test classes from the botocore test classes with some meta class.
something like starting from a common (maybe abstract) base class for the botocore tests, finding the actual botocore test classes from that, and use the meta class to generate an adapted test class for aiobotocore.
the base class would probably need some factory method for creating the botocore client/resource instances and for the aiobotocore instances some wrapping to make the async functions callable like sync functions.

This is just some wild idea and i'm not sure if it will even work, like:
* will the dynamically created tests be detected by the test framework
* are the botocore tests even installed and accessible when installing the aiobotocore instrumentation for testing
* possible nuances in the handling between botocore and aiobotocore which would require different handling in certain test cases
* ...

so i'm not sure if it would be really worth to try out this approach, also since it most likely will make the test code hard to follow and less readable.

duplicating the tests is also fine i gues and if they really get out of sync it shouldn't be a big problem since the AioBotocoreInstrumentor more or less just calls into and reuses the botocore instrumentation.

**Review by mariojonke — APPROVED** [review]

File: instrumentation/opentelemetry-instrumentation-aiobotocore/tests/test_aiobotocore_dynamodb.py
Comment (dacevedo12) [reply instrumentation/opentelemetry-instrumentation-aiobotocore/tests/test_aiobotocore_dynamodb.py:1]: I tried, but couldn't import them as the tests are not included in the library... doing some sort of client proxy that async-ifies `client.something` calls would be possible, but I share the concern that it would affect readability.

Something else to notice though is that aiobotocore will almost always use an older version of botocore underneath, as the patching efforts are a manual process, and often [not trivial](https://github.com/aio-libs/aiobotocore/pull/999#issuecomment-1457485410).

**Comment by nemoshlag** [comment]
Hi @dacevedo12 thanks but it's ok. I think new instrumentations should be promoted, but I don't have any control over merging prioritization :)

**Comment by dacevedo12** [comment]
Perhaps the bot requested more reviews due to the spellcheck fix in boto3sqs. I just changed it so the CI would pass

Any chance this makes it to the 0.38 release :smile: ?

**Comment by ronlut** [comment]
@dacevedo12 thanks for your precious time and work :)
I am waiting for this as well, is it still blocked? Who should approve it?
Maybe worth releasing it as a separate instrumentor if it's not gonna happen in this repo?

**Comment by dacevedo12** [comment]
They likely won't merge it, check the (ongoing?) discussion at #1714.

I'm down to publish this on a separate repo, but won't happen immediately as I'm busy some stuff.

Feel free to fork/copypaste in the meantime, that's what I did at work.

**Comment by dacevedo12** [comment]
Will move this to my own repo soon. Progress surely won't be made here

**Comment by idonava** [comment]
Is there any chance to merge this PR?
This library is widely used, and we were also waiting for it.

**Comment by henocdz** [comment]
@lzchen @ocelotl I see you have been working on reviewing and approving recent MRs on this repo; do you know what's required to get this much-needed change? How can we help? It'd be nice to have this available. 

Thanks for your work maintaining this.

**Pull request closed without merging**
